agent/cache/watch.go

```diff
@@ -0,0 +1,103 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+	"time"
+)
+
+// WatchUpdate is a struct summarising an update to a cache entry
+type WatchUpdate struct {
+	Result interface{}
+	Meta   ResultMeta
+	Err    error
+}
+
+// indexManagedRequest is a wrapper around a user-passed Request that allows us
+// to manage the blocking index without knowing how to modify the concrete
+// Request type.
+type indexManagedRequest struct {
+	r        Request
+	minIndex uint64
+}
```

Comment:
@mitchellh FYI, this turns out to be a big problem with this design...

All the tests with mock cache types work wonderfully, but when I'd got all the way to using this for real somewhere, the real cache types all type-check the request object and error because they get an `indexManagedRequest` and not the type they wanted 😢 .

I'll re-think how we can abstract managing the query options on request types without wrapping the request object I guess.